### ISSUE TITLE
Proposal: tsc backed jsx conditional rendering rule

### ISSUE THREAD

**Issue by pke** [typescript-eslint/typescript-eslint#2770]
Over at https://github.com/yannickcr/eslint-plugin-react/issues/2073 we are discussing a rule that only the TSC could help with.
It would require to detect if a conditional rendering check for react JSX blocks of code is a true boolean value.

Do you think the JSX world and this eslint plugin could somehow be combined to create a rule (in a new package?) for this particular bad coding error (that will crash any react-native app)?

**Comment by bradzacher** [comment]
It seems like you just want [`strict-boolean-expressions`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/strict-boolean-expressions.md) to also check JSX expressions as if they were logical conditions?

**Comment by pke** [comment]
I think that would do it, yes.

**Comment by bradzacher** [comment]
happy to accept PR for a new option, say `checkJsxLogicals` which applies the rule's logic to a JSX expression (if it is a logical expression).

**Comment by pke** [comment]
I have [written](https://github.com/pke/eslint-plugin-redux-saga) eslint rules before so I am somewhat familiar with the concept. Is there any JSX support yet in this plugin I could use as a starting point?

**Comment by marhaupe** [comment]
I'm just chiming in to say that I'd be happy to test this once you've implemented this option.

**Comment by bradzacher** [comment]
Best option would be to play around in https://astexplorer.net/ to get a feel for the AST you're looking to target.

From there you can craft an [ESLint selector](https://eslint.org/docs/developer-guide/selectors) as best you can to target that AST.

Then you need to [work it into the rule](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/rules/strict-boolean-expressions.ts#L134-L142).

If you want a live playground for inspecting the type information that TS provides - https://ts-ast-viewer.com/ is an amazing site.

**Comment by phaux** [comment]
[`strict-boolean-expressions`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/strict-boolean-expressions.md) already does that and I use it for this exact purpose. You just have to set `allowNumber` to `false` (it's `true` by default) and you get:

```tsx
<div>
  {items.length && (
    // ^ Unexpected number value in conditional. An explicit zero/NaN check is required.
    <ul>
      {items.map(item => <li>{item}</li>)}
    </ul>
  )}
</div>
```

The only possible downside is that if you disable `allowNumber` then it's disallowed everywhere, not just in JSX, so you have to write e.g. `if (items.length > 0)` instead of just `if (items.length)` (which is fine for me).

**Comment by bradzacher** [comment]
that is true - it will "just work" with the right options.
I wonder if this is just a documentation thing, then?

We can add another section to the docs that say "if you want to strictly validate your JSX logical expressions do not output values, you should use this config"

**Comment by phaux** [comment]
@bradzacher Yep. Let's add a paragraph to allowNumber and allowString options in docs that says it's useful to set them to false in react projects. I can send a PR.

It's not a problem specific to JSX though. It applies everywhere where any value can be passed. 
E.g. ``console.log(items.length && `There are ${items.length} items.`)`` would log the message about length or just "0".

**Comment by phaux** [comment]
This all comes down to how logical operators are used and how you sometimes use them only for their short-circuiting behavior.
Let's say you disable allowNullableObject and then suddenly you can't do `const item = itemMap.get("id") || defaultItem` (yes, you simply use nullish coalescing now and it works but it's just an example).

I remember that I have already suggested to have all the `allow*` options accept 3 options. E.g:
- `true` or `"always"` – as it is now
- `"inExpression"` or something like that – only when using logical operator for it's side effect. (can be checked if AST parent is not if/for/while etc)
- `false` or `"never"` – same as now

but is it worth it to maintain such a complicated rule? I don't know.

Another idea I had in mind would be to forbid `a && b` and `a || b` altogether when they're not in condition, but that would fit better as a rule in the original eslint project.

These are just my random thoughts. I'm not proposing any of these.

**Comment by hpersson** [comment]
cool, I took a stab at implementing the rule using `@typescript-eslint/experimental-utils` and added it to our project - 
[eslint-plugin-jsx-expressions](https://github.com/hpersson/eslint-plugin-jsx-expressions/) 
just in case anyone else discovers this thread and finds it useful!

**Comment by ypresto** [comment]
@hpersson This issue was marked as accepting PRs. Could you send PR to include your rule in this repo?

**Comment by pedrodurek** [comment]
Any updates?

**Comment by phaux** [comment]
How about a rule that is like [`restrict-template-expressions`](https://typescript-eslint.io/rules/restrict-template-expressions) but it checks JSX instead of template strings:

```jsx
return (
  <p>Hello {name}</p> // report if name is not string
)
return (
  <p>Items: {items.length && items.join(", ")}</p> // error: type number is not allowed in JSX expression
)
```

It would also be useful to not forget to format numbers using user's language:

```js
// bad:
function ScoreCounter(props) {
  return (
    <p>{props.score} 🪙</p>
  )
}

// good:
import { FormattedNumber } from "react-intl"
function ScoreCounter(props) {
  return (
    <p><FormattedNumber value={props.score} /> 🪙</p>
  )
}
```

I can send PR.

**Comment by pedrodurek** [comment]
In my view, for JSX conditional rule, we don't need any configurable options, it should only accept booleans and nulls, excluding any other falsy and truthy values. That way we can tackle all edge cases for React and React Native apps.

**Comment by karlhorky** [comment]
> To be clear - if we want to make a small enhancement to `strict-boolean-expressions` to do an additional check for JSXExpressions - then that change lives here in this project.

Funny enough, I'm actually coming back here after using @hluisson's `eslint-plugin-jsx-expressions` for a long time, which has worked well! But [no longer supports latest `typescript-eslint` or latest ESLint](https://github.com/hluisson/eslint-plugin-jsx-expressions/pull/17).

So I proposed to @JoshuaKGoldberg to open a new issue about only JSX, to absorb the functionality of `eslint-plugin-jsx-expressions` into upstream again.

I was thinking about something like:

```json
{
  "rules": {
    "@typescript-eslint/strict-boolean-expressions": ["error", {
      "allowString": true,
      "allowNumber": true,

      "jsxExpressions": {
        "allowString": false,
        "allowNumber": false
      }
    }]
  }
}
```

This would report for `num` and `str` in the 2 JSX expressions, but \*would not\* report for `num2` and `str2`:

```tsx
let num = 0;
let str = '';

<>
  <div>{num && 'x'}</div>
  <div>{str && 'x'}</div>
</>

let num2 = 0;
let str2 = '';

if (num2 && str2) { }
```

**Comment by bradzacher** [comment]
Personally I'd have thought that we wouldn't even need an option.
Really we just want to treat JSX expression containers as all other logical containers (I.e same as `if` conditions, etc).

Maybe one option to toggle the JSX handling on or off, but I don't think individual type options would be correct - but I don't rwally see a reason that you wouldn't want this behaviour if you're using the rule.

**Comment by JoshuaKGoldberg** [comment]
Yeah the tricky things here are:

* JSX is just a syntax, not a runtime behavior spec. Different frameworks built around the syntax can treat it differently.
* There are more places than just a `children` prop that common frameworks such as Solid and React allow.

On the latter note, suppose you switch from a `children`-based component:

```tsx
declare const MyComponent: React.FC<React.PropsWithChildren>;
<MyComponent>{str && 'x'}</MyComponent>
//            ~~~~~~~~~~
```

...to one with >=1 render prop:

```tsx
declare const MyComponent: React.FC<{
    first: React.ReactNode;
    second: React.ReactNode;
}>;
declare const str: boolean;

<MyComponent
    first={str && 'x'}
    second={str && 'x'}
/>
```

That's a common thing devs do, including using `React.ReactNode` rather than the more precise `string | React.JSX.Element`.

I still think it'd be better to have nuanced JSX best practices exist in a JSX-dedicated plugin. We don't have the ability to focus so granularly here on it. 

In other words: I think _this_ issue for typescript-eslint is still a docs one.

**Comment by JoshuaKGoldberg** [comment]
Cool! So thinking again on the requested functionality: it's framework-specific. JSX allows you to put whatever child you want in your components. The fact that you might not want to render a `0`, `""`, `null`, etc. is framework-specific. +1 to plugins such as the new [`eslint-plugin-react-x`](https://www.npmjs.com/package/eslint-plugin-react-x) ([GitHub](https://github.com/Rel1cx/eslint-react/tree/main/packages/plugins/eslint-plugin-react-x), [docs](https://eslint-react.xyz/)) building their own answers here.

Closing out as out-of-scope for typescript-eslint.

To anybody wondering _"but wait, I need this!"_ -- yes,  agreed, it'd be a really useful rule! https://typescript-eslint.io/developers/custom-rules has docs for how you could build this yourself to your UI framework's needs.

**Comment by karlhorky** [comment]
I switched successfully from `eslint-plugin-jsx-expressions` to `react-x/no-leaked-conditional-rendering`, tested and confirmed to be working 🎉 

If you want to see the PR and diff:

- https://github.com/upleveled/eslint-config-upleveled/pull/424
